### PR TITLE
Index blocks by number in the DB

### PR DESCRIPTION
**Pull request by snd**
#780 

- [x] make tests pass
- [x] ~~test reorgs~~
- [x] rebase
- [x] adapt core/client/db/src/cache to changes
- [x] make tests pass after rebase
- [x] db/light: fix pruning on `note_finalized`

### CODE REVIEW

File: core/client/db/src/light.rs [hunk 1]
```diff
@@ -263,27 +266,55 @@ impl<Block> LightBlockchainStorage<Block> for LightStorage<Block>
 								(&retracted.number, &retracted.hash));
 						}
 
-						transaction.delete(
-							columns::HASH_LOOKUP,
-							&::utils::number_to_lookup_key(retracted.number)
-						);
+						let prev_lookup_key = ::utils::number_to_lookup_key(retracted.number);
```
Comment (pepyakin) [inline core/client/db/src/light.rs]: This code is pretty wide (up to whooping 178 chars). Can we do something with it?

File: core/client/db/src/lib.rs [hunk 2]
```diff
@@ -569,7 +564,7 @@ impl<Block> client::backend::Backend<Block, Blake2Hasher> for Backend<Block> whe
 						BlockId::Hash(parent_hash),
 					)?;
 
-					// update block number to hash lookup entries.
+					// uncanonicalize
```
Comment (svyatonik) [inline core/client/db/src/lib.rs]: Probably extract `for` in a separate `fn`? Seems like difference of for-s is only the `prev_lookup_key` && `new_lookup_key` + returning `NotInFinalizedChain` err in `retracted` for?

File: core/client/db/src/utils.rs [hunk 3]
```diff
@@ -82,6 +83,49 @@ pub fn number_to_lookup_key<N>(n: N) -> BlockLookupKey where N: As<u64> {
...
+		| (key[1] as u64) << 16
+		| (key[2] as u64) << 8
+		| (key[3] as u64)).map(As::sa)
+}
+
+/// Convert block id to block lookup key.
+/// block lookup key is the DB-key header, block and justification are stored under.
+/// looks up lookup key by hash from DB as necessary.
+pub fn block_id_to_lookup_key<Block>(
+    db: &KeyValueDB,
```
Comment (svyatonik) [inline core/client/db/src/utils.rs]: spaces -> tabs

File: core/client/db/src/lib.rs (hunk 2 above)
Comment (gavofyork) [reply core/client/db/src/lib.rs]: @snd maybe address this in a followup PR?

**Pull request merged**